samplekotlin/src/main/java/com/example/chattutorial/ChannelActivity4.kt

```diff
@@ -51,62 +51,51 @@ class ChannelActivity4 : AppCompatActivity(R.layout.activity_channel_4) {
             }
         }
-        // step 4 - let the message input know when we are editing a message
-        messageListView.setOnMessageEditHandler {
-            messageInputViewModel.editMessage.postValue(it)
-        }
 
-        // custom typing info header bar
-        val channelController = ChatClient.instance().channel(cid)
-        val currentlyTyping = MutableLiveData<Set<String>>(emptySet())
-
-        channelController.subscribeFor(TypingStartEvent::class.java, TypingStopEvent::class.java) {
-            val typing = currentlyTyping.value ?: emptySet()
-            val typingCopy = when (it) {
-                is TypingStartEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typing.toMutableSet() + name
-                }
-                is TypingStopEvent -> {
-                    val name = it.user.extraData["name"] as String
-                    typing.toMutableSet() - name
-                }
-                else -> {emptySet()}
+        messageListViewModel.state.observe(this) {
+            when (it) {
+                is MessageListViewModel.State.NavigateUp -> finish()
             }
-            currentlyTyping.postValue(typingCopy)
         }
 
-        val typingObserver = Observer<Set<String>> { users ->
-            channelHeaderSub.text = if (users.isEmpty()) {
-                "nobody is typing"
-            } else {
-                "typing: " + users.joinToString(", ")
-            }
+        // step 4 - let the message input know when we are editing a message
+        messageListView.setOnMessageEditHandler { message ->
+            messageInputViewModel.editMessage.postValue(message)
         }
-        currentlyTyping.observe(this, typingObserver)
+
+        val currentlyTyping = mutableSetOf<String>()
 
         // step 5 - handle back button behaviour correctly when you're in a thread
-        val backButtonHandler = {
+        channelHeaderView.onBackClick = {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
         }
-        channelHeaderView.onBackClick = { backButtonHandler() }
 
-        onBackPressedDispatcher.addCallback(
-            this,
-            object : OnBackPressedCallback(true) {
-                override fun handleOnBackPressed() {
-                    backButtonHandler()
+        // make pressing back on hardware the same as channel header's back
+        onBackPressedDispatcher.addCallback(this) {
+            channelHeaderView.onBackClick()
+        }
+
+        // custom typing info header bar
+        ChatClient
+            .instance()
+            .channel(cid)
+            .subscribeFor(TypingStartEvent::class.java, TypingStopEvent::class.java) { event ->
+                when (event) {
+                    is TypingStartEvent -> currentlyTyping.add(event.user.name)
+                    is TypingStopEvent -> currentlyTyping.remove(event.user.name)
+                }
+
```

Comment:
this callback isn't lifecycle aware and will give errors when the activity is gone (and the callback is still running.)
probably fine for a tutorial/example... @JcMinarro what do you think?

Reply:
IMO we should fix that. Most of the users will just copy-paste things from a tutorial and that's why we should show how to do everything correctly

Reply:
What do you think about 
```kotlin
public fun ChannelController.subscribeFor(
    vararg eventTypes: Class<out ChatEvent>
): LiveData<ChatEvent> = MutableLiveData<ChatEvent>().also { eventsLiveData ->
    subscribeFor(*eventTypes) { event ->
        eventsLiveData.postValue(event)
    }
}

...

        ChatClient
            .instance()
            .channel(cid)
            .subscribeFor(TypingStartEvent::class.java, TypingStopEvent::class.java)
            .observe(this) { event ->
                when (event) {
                    is TypingStartEvent -> currentlyTyping.add(event.user.name)
                    is TypingStopEvent -> currentlyTyping.remove(event.user.name)
                }

                typingHeader.text = when {
                    currentlyTyping.isNotEmpty() -> currentlyTyping.joinToString(prefix = "typing: ")
                    else -> nobodyTyping
                }
            }
```

Reply:
The above option doesn't call `dispose()` on the `Disposable` we get from `subscribeFor`; LiveData just ignores values pushed to it. If we care about calling dispose, maybe we could do something like
```kotlin
public fun ChannelController.subscribeFor(
    lifecycleOwner: LifecycleOwner,
    vararg eventTypes: Class<out ChatEvent>,
    listener: (ChatEvent) -> Unit,
) {
    subscribeFor(*eventTypes) { event ->
        listener(event)
    }.also { subscription ->
        lifecycleOwner.lifecycle.addObserver(
            object : LifecycleObserver {
                @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
                fun onDestroy() {
                    if (subscription.isDisposed) {
                        return
                    }

                    subscription.dispose()
                }
            }
        )
    }
}
```